### PR TITLE
26 minimum viable product create template page for challenge creation

### PR DESCRIPTION
**Pull request by jaidan18**
Created challenge creation page with its associated flask form and route, following the styling of the login and challenge view pages.

challengecreationpage2

### CODE REVIEW

File: app/__init__.py [hunk 1]
```diff
@@ -1,5 +1,7 @@
 from flask import Flask
+from app.config import Config
```
Comment (RedBlueCarrots) [inline app/__init__.py:2]: Config things will need to be sorted in a merge - I think keep whatever is currently in the main branch, assuming that still works for you

File: app/config.py [hunk 2]
```diff
@@ -0,0 +1,5 @@
+import os
+
+class Config:
```
Comment (RedBlueCarrots) [inline app/config.py:3]: likewise, config stuff should be in main already - a local merge will fix this

File: app/forms.py [hunk 3]
```diff
@@ -0,0 +1,11 @@
+from flask_wtf import FlaskForm
+from wtforms import StringField, IntegerField, SubmitField, DateTimeLocalField
+from wtforms.validators import DataRequired
+
+class ChallengeCreationForm(FlaskForm):
```
Comment (RedBlueCarrots) [inline app/forms.py:5]: looks great!

File: app/templates/create.html [hunk 4]
```diff
@@ -0,0 +1,63 @@
+{% extends "base.html" %}
+
+{%block content%}
+<div class="container-fluid  min-vh-100 bg-light">
+  <h1 class="my-3 text-center">Create a Challenge</h1>
+  
+  <form class="container my-1 rounded bg-white border border-primary" id="createForm" data-link="{{ url_for('create') }}">
```
Comment (RedBlueCarrots) [inline app/templates/create.html:7]: just as a heads up - most likely for the next milestone we will want to use ajax for this, but it works for the MVP and is much simpler

File: app/templates/create.html [hunk 5]
```diff
@@ -0,0 +1,63 @@
...
+    {{ form.hidden_tag() }}
+
+    <div class="row my-3 gy-2">
+      {{ form.title.label(class="col-form-label col-2") }}
+      <div class="col-10">
+        {{ form.title(class="form-control") }}
```
Comment (RedBlueCarrots) [inline app/templates/create.html:13]: If I understand this correctly, submitting an invalid form will refresh the page and load in the errors when it refreshes?

**Review by RedBlueCarrots — COMMENTED** [review]
Will approve once a local merge with main has been done

File: app/forms.py [hunk 6]
```diff
@@ -0,0 +1,11 @@
...
+    title = StringField('Challenge Title', validators=[DataRequired()])
+    start = StringField('Starting Article', validators=[DataRequired()])
+    destination = StringField('Destination Article', validators=[DataRequired()])
+    time = DateTimeLocalField('Submission Close', validators=[DataRequired()])
+    attempts = IntegerField('Submission Attempts', validators=[DataRequired()])
```
Comment (DeltaO3) [inline app/forms.py:10]: Is this field necessary (attempts) ? I dont think it is doing anything

Comment (DeltaO3) [reply app/forms.py:10]: Also, please do a local merge, as this forms.py already should exist from an earlier PR

File: app/templates/create.html (hunk 4 above)
Comment (DeltaO3) [reply app/templates/create.html:7]: if its a seperate page ajax wont be necessary, wtforms does everything needed and displays rerrors on re render of a page.

File: app/templates/create.html (hunk 5 above)
Comment (DeltaO3) [reply app/templates/create.html:13]: Yes, It should to me understanding and research (if youre curious re-read the server side flask lecture, and do part 3 of the flask mega tutorial, where this is covered:) )

File: app/templates/create.html [hunk 7]
```diff
@@ -0,0 +1,63 @@
...
+{%block content%}
+<div class="container-fluid  min-vh-100 bg-light">
+  <h1 class="my-3 text-center">Create a Challenge</h1>
+  
+  <form class="container my-1 rounded bg-white border border-primary" id="createForm" data-link="{{ url_for('create') }}">
+    {{ form.hidden_tag() }}
+
+    <div class="row my-3 gy-2">
+      {{ form.title.label(class="col-form-label col-2") }}
+      <div class="col-10">
```
Comment (DeltaO3) [inline app/templates/create.html:12]: The display breaks a bit on mobile - use col-sm-10 instead? 
image

File: app/templates/create.html [hunk 8]
```diff
@@ -0,0 +1,63 @@
...
+        {{ form.attempts(class="form-control") }}
+      </div>
+      {% for error in form.attempts.errors %}
+      <span style="color: red;">[{{ error }}]</span>
+      {% endfor %}
+    </div>
+
+    <div class="my-3">{{ form.submit(class="btn btn-primary") }}</div>
+  </form>
```
Comment (DeltaO3) [inline app/templates/create.html:61]: Part of me feels like for such a big repeat of mildly edited code there *should* be a jinja way to put it in afor loop using the form class. Not necessary for this issue, but potentially one that could be looked into

**Review by DeltaO3 — CHANGES_REQUESTED** [review]
Looks ok, but theres definitely an issue with mobile formatting (I believe you just forgot to make col-sm and just used col). Also, as joseph said, you need to update from main.

File: app/templates/create.html (hunk 8 above)
Comment (jaidan18) [reply app/templates/create.html:25]: I looked into it and there is. You can simply loop through the fields of the form.  I have now made this change to the challenge creation template.

File: app/templates/create.html (hunk 7 above)
Comment (jaidan18) [reply app/templates/create.html]: I've changed all of the col-10 to col-sm-10 and I believe this has fixed the problem.

File: app/config.py (hunk 2 above)
Comment (jaidan18) [reply app/config.py:3]: I've done a local merge and it resolved most of the merge conflicts, there's just a few easy-to-fix conflicts left now.

File: app/templates/base.html [hunk 9]
```diff
@@ -5,10 +5,45 @@
...
+				aria-label="Toggle navigation">
+				<span class="navbar-toggler-icon"></span>
+			</button>
+			<div class="collapse navbar-collapse" id="navbarSupportedContent">
+				<ul class="nav nav-pills nav-fill me-auto mb-2 mb-lg-0">
+					<li class="nav-item">
+						<a class="nav-link active" href="#">Play</a>
+					</li>
+					<li class="nav-item">
+						<a class="nav-link" href="#">Create</a>
```
Comment (DeltaO3) [inline app/templates/base.html:28]: href should be {{url_for(create)}} so we can actually navigate to the create page

File: app/templates/create.html [hunk 10]
```diff
@@ -0,0 +1,27 @@
...
+<div class="container-fluid  min-vh-100 bg-light">
+  <h1 class="my-3 text-center">Create a Challenge</h1>
+  
+  <form class="container my-1 rounded bg-white border border-primary" id="createForm" method="put">
+    {{ create_form.hidden_tag() }}
+
+    {% for field in create_form %}
+    {% if field.type == "StringField" or field.type == "DateTimeLocalField" %}
+    <div class="row my-3 gy-2">
+      {{ field.label(class="col-form-label col-2") }}
```
Comment (DeltaO3) [inline app/templates/create.html:13]: Formatting still weird on mobile: col 2 is there regardless of size still (use either col-sm-2 or just col).
Consider for such a big form only using margin top - at least personally i feel theres too much space between input elements.

**Review by DeltaO3 — CHANGES_REQUESTED** [review]
Looking good, but theres still some minor formatting error on mobile (look for yourself, but the label curently takes up two line on mobile where it could easily just  take up one.), and you could add the routing to the create page from the navbar

File: app/routes.py [hunk 11]
```diff
@@ -18,11 +18,12 @@ def index():
 #Create challenge page
 @app.route('/create', methods=['GET', 'PUT'])
 def create():
-    #TODO
+    form = LoginForm()
+    create_form = ChallengeCreationForm()
     if request.method == "PUT":
```
Comment (DeltaO3) [inline app/routes.py:23]: this line is actually unnecessary for now believe, flask will check the next if statement anyway

File: app/routes.py [hunk 12]
```diff
@@ -18,11 +18,12 @@ def index():
 #Create challenge page
 @app.route('/create', methods=['GET', 'PUT'])
```
Comment (DeltaO3) [inline app/routes.py:19]: PUT should be POST

File: app/templates/create.html [hunk 13]
```diff
@@ -0,0 +1,27 @@
+{% extends "base.html" %}
+
+{%block content%}
+<div class="container-fluid  min-vh-100 bg-light">
+  <h1 class="my-3 text-center">Create a Challenge</h1>
+  
+  <form class="container my-1 challenge-border bg-white" id="createForm" method="put">
```
Comment (DeltaO3) [inline app/templates/create.html:7]: method should be POST, not put (i think case matters)

File: app/templates/create.html [hunk 14]
```diff
@@ -0,0 +1,27 @@
+{% extends "base.html" %}
+
+{%block content%}
+<div class="container-fluid  min-vh-100 bg-light">
+  <h1 class="my-3 text-center">Create a Challenge</h1>
```
Comment (DeltaO3) [inline app/templates/create.html:5]: my-3 is creating the gap - but also maybe consider removing the background, and moving this header *inside* the box as per our mockup?

**Review by DeltaO3 — COMMENTED** [review]
Looking much better, just a few more minor tidbits and ill be able to approve this :)

File: app/templates/create.html (hunk 14 above)
Comment (DeltaO3) [reply app/templates/create.html:5]: (im actually unsure if my is creating the gap but i dont see what else would be. either way, you should move inside the box.)

File: app/templates/create.html (hunk 13 above)
Comment (jaidan18) [reply app/templates/create.html:7]: I've changed it to post now, but still in lowercase, as this is the convention in HTML5.

**Review by DeltaO3 — APPROVED** [review]
LGTM! 

You might wanna look into your git workflow though, you seem to have done something to make all the commits happen within 2 hours somehow (its fine tho)

**Pull request merged**